torch/csrc/autograd/variable.cpp

```diff
@@ -151,51 +78,47 @@ void Variable::backward(
 
   std::vector<Variable> inputs;
   if (!gradient.has_value()) {
-    gradient = make_variable(at::ones_like(data()), /*requires_grad=*/false);
+    gradient = make_variable(at::ones_like(*this), /*requires_grad=*/false);
   }
   inputs.push_back(std::move(as_variable_ref(*gradient)));
   Engine::get_default_engine().execute(edges, inputs, keep_graph, create_graph);
 }
 
-void Variable::Impl::set_data(Tensor new_data) {
+void Variable::set_data(Tensor new_data) const {
+  AT_ASSERT(is_sparse() == new_data.is_sparse());
```

Comment:
In the old `variable.set_data(data_tensor)` method, it was ok to pass a sparse `data_tensor` to a dense `variable` (or vice versa). In the new `variable.set_data(data_tensor)` method, this behavior is not possible to implement (we can't just swap out the variable's TensorImpl and put in a SparseTensorImpl), we instead choose to disallow this behavior, and fix places where this behavior is expected.

Reply:
Is there a way to generalize this test so that it's not hardcoded for sparse? I'm not even sure what the right test should be. Maybe, TensorTypeId has to be consistent? (It shouldn't include dtype, since that's explicitly permitted in the code below.)